### ISSUE TITLE
Construction and assignment from lvalue require exact type match

### ISSUE THREAD

**Issue by pbackus** [pbackus/sumtype#47]
From <https://forum.dlang.org/thread/eiepxeebjxljsoquddyb@forum.dlang.org#post-pnufczmiecexaicklnrw:40forum.dlang.org>:

Older version of sumtype accept this code:

```d
void main(){
    import sumtype;
    alias Val = SumType!(bool);

    const bool b = true;
    Val val = b;  //fail in newest version
    val = b;      //fail in newest version
}
```

but new version need exact type:

```d
void main(){
    import sumtype;
    alias Val = SumType!(bool);

    bool b = true;   //cannot be const
    Val val = b;
    val = b;
}
```

**Comment by pbackus** [comment]
The proximate cause of this issue is that `SumType`'s constructors take lvalues by `ref`, and arguments passed to `ref` parameters must match the parameter's type exactly.

To fix this, it would be necessary to change the signature of `SumType`'s constructor such that it accepts any type which implicitly converts to the appropriate member type. For example:

```d
this(Value)(auto ref Value value)
    if (is(Value : T))
```

However, making this change introduces a new problem: because the parameter type of the instantiated constructor will *always* be an exact match for the argument type, we can no longer rely on [partial ordering](https://digitalmars.com/articles/b20.html) to choose the correct constructor overload in cases where one member type is implicitly convertible to another (e.g., `SumType!(int, double)`).

Unlike the existing issue, this new problem cannot be worked around in client code. Therefore, this issue will not be fixed.

**Comment by pbackus** [comment]
Found a potential workaround for this, but I'd like to check the generated assembly to make sure its performance is acceptable before merging.

**Comment by FeepingCreature** [comment]
Just ran into this. It's a crippling issue if you want to use const, approximately ever.

**Comment by pbackus** [comment]
Fixed and tagged as `v1.0.3`.
